### PR TITLE
suppress output from intentionally failing tests

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes https://github.com/flutter/flutter/issues/82873

### CODE REVIEW

**Comment by Hixie** [comment]
This seems really complicated, can't we just redirect to /dev/null or something?

**Comment by yjbanov** [comment]
> This seems really complicated, can't we just redirect to /dev/null or something?

I'm not sure. Is there a way to redirect `dart:io/stdout` to /dev/null? The issue is that `package:test_core` writes to stdout directly, so I needed to somehow convince it to stop doing that. Also, wouldn't redirecting `stdout` shut off output globally? Here I just want to shut off output from the test, but keep others, such as the line that says "Running some_test.dart".

FWIW I filed https://github.com/dart-lang/test/issues/1521 and left a TODO here. Once it's fixed, all this becomes a one-liner.

**Comment by yjbanov** [comment]
Unless we split this code from `felt.dart` into a separate Dart program and call it as a child process instead of calling it as a function. Then we can redirect the child process' stdout to /dev/null. Without trying it's hard to say if that would be net plus or minus in complexity.

**Comment by Hixie** [comment]
The equivalent code in the framework was originally just a few lines of bash, FWIW. I think now it's a few lines in the test.dart, though that code itself is probably more complicated than strictly necessary at this point and could benefit from a cleaning where we remove the non-critical features that we've accrued over time. A thousand lines just to bootstrap the tests seems like a lot of code to me.

**Comment by yjbanov** [comment]
I agree this file is a mess. We got here after a long chain of small incremental feature additions. It needs a major clean up. This file is responsible for computing a subset of tests to run (a magical and my least favorite feature; I think we can just remove it), compiling tests, handle goldens and test fonts, and running tests. We can clean this up by separating these concerns.

I don't expect a significant reduction in the amount of code though. It will just be cleaner. The equivalent code on the flutter side is `dev/bots/test.dart` plus a few files in https://github.com/flutter/flutter/tree/master/packages/flutter_tools/lib/src/test. So, probably 1-2KLoC too.

Let me know how you'd like to proceed. If silencing failing test outputs is not urgent, I can do the clean-up first.

**Comment by Hixie** [comment]
If someone more familiar with the code is able to review it I'm certainly not going to stand in the way. :-) I don't think I understand these APIs well enough to know if this is a good change (certainly it's concerning to see the PR itself describe this as a hack).

**Comment by yjbanov** [comment]
> If someone more familiar with the code is able to review it I'm certainly not going to stand in the way. :-) I don't think I understand these APIs well enough to know if this is a good change (certainly it's concerning to see the PR itself describe this as a hack).

No worries. I'll find someone to review. As for the hack, this is continuing the [existing tradition](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/test/test_wrapper.dart#L8).

File: lib/web_ui/dev/test_runner.dart [hunk 1]
```diff
@@ -850,3 +876,23 @@ class TestBuildInput {
...
 
 class TestFailureException implements Exception {}
+
+/// Prints the name of the test, but suppresses all other test output.
+///
+/// This is useful to prevent pollution of logs by tests that are expected to
+/// fail.
+class NameOnlyReporter implements hack.Reporter {
+  NameOnlyReporter(hack.Engine testEngine) {
+    testEngine.onTestStarted.listen(_printTestName);
```
Comment (mdebbar) [inline lib/web_ui/dev/test_runner.dart:886]: I'm not sure if we need this, but we probably do:
```suggestion
    final sub = testEngine.onTestStarted.listen(_printTestName);
    testEngine.success.then((_) {
      sub.cancel();
    });
```

File: lib/web_ui/dev/test_runner.dart [hunk 2]
```diff
@@ -850,3 +876,23 @@ class TestBuildInput {
...
+
+  void _printTestName(hack.LiveTest test) {
+    print('Running ${test.groups.map((hack.Group group) => group.name).join(' ')} ${test.individualName}');
+  }
+
+  @override
+  void pause() {}
+
+  @override
+  void resume() {}
```
Comment (mdebbar) [inline lib/web_ui/dev/test_runner.dart:897]: We should respect `pause` and `resume`.
```suggestion
  bool _isPaused = false;

  void _printTestName(hack.LiveTest test) {
    if (_isPaused) {
      return;
    }
    print('Running ${test.groups.map((hack.Group group) => group.name).join(' ')} ${test.individualName}');
  }

  @override
  void pause() {
    _isPaused = true;
  }

  @override
  void resume() {
    _isPaused = false;
  }
```

Comment (yjbanov) [reply lib/web_ui/dev/test_runner.dart:900]: This reporter is effectively always paused by never printing anything, so I think that would be superfluous. I believe test name is printed before the test is paused. But if you think this is useful, I'll take the suggestion.

**Review by mdebbar — APPROVED** [review]

**Pull request merged**